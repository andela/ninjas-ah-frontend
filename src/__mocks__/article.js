export default (Object.article = {
  title: 'my title well done',
  description: 'description of the article: well',
  body: 'body of the article: well'
});
