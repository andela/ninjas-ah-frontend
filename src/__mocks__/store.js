import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import initialState from '../store/initialState';
import apiMiddleware from '../middlewares/apiMiddleware';

export { initialState };
export const middlewares = [thunk, apiMiddleware];
export const mockStore = configureMockStore(middlewares);

export default mockStore(initialState);
