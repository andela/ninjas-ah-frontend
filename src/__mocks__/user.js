export default (Object.user = {
  firstName: 'John',
  lastName: 'Smith',
  username: 'josmi',
  email: 'josmi@email.com',
  image: 'image.jpg'
});

export const userToRegister = {
  ...Object.user,
  password: 'Abcd1234!',
  confirmPassword: 'Abcd1234!'
};

export const mismatchedUserPassword = {
  ...Object.user,
  password: '12345678',
  confirmPassword: '1234abcd'
};

export const wrongUserPassword = {
  ...Object.user,
  password: 'Abcd1234'
};

export const resetPassword = {
  password: '12345678',
  confirmPassword: '12345678'
};

export const mismatchedResetPassword = {
  password: '12345678',
  confirmPassword: '1234abcd'
};

export const matchedResetPassword = {
  passwordOne: 'Brazzaville10!',
  passwordTwo: 'Brazzaville10!'
};

export const sendEmail = { email: 'noreplay@gmail.com' };
export const fakeEmail = { email: 'fsdfsdfsdfdsf--gmail.com' };
