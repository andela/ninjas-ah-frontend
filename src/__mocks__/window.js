const map = {};
window.document.addEventListener = jest.fn((event, callback) => {
  map[event] = callback;
});
window.addEventListener = jest.fn((event, callback) => {
  map[event] = callback;
});
window.getSelection = jest.fn(() => ({
  toString: jest.fn(() => 'highlightedText'),
  getRangeAt: jest.fn(() => ({
    getBoundingClientRect: jest.fn(() => ({
      top: 0,
      bottom: 0,
      left: 0,
      right: 0
    }))
  }))
}));

window.document.querySelectorAll = jest.fn(() => [
  {
    removeEventListener: jest.fn((event, callback) => {
      map[event] = callback;
    }),
    addEventListener: jest.fn((event, callback) => {
      map[event] = callback;
    })
  }
]);

window.scrollTo = jest.fn(() => 1);

const event = jest.fn(e => map[e.name](e));
export const document = { event };
export default { ...document, document };
