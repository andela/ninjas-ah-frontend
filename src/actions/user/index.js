import * as getUser from './getUser';
import signup from './signup';
import socialAuth from './socialAuth';

export { getUser, signup, socialAuth };
