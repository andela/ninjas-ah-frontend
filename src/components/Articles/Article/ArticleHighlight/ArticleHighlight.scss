@import '../../../../assets/css/_colors';
@import '../../../../assets/css/_radius';
@import '../../../../assets/css/_shadow';

.ArticleHighlight {
  .selection {
    @extend .radius-2;
    text-align: center;
    background: lightgreen !important;
    padding: 5px;
    font-style: italic;
    font-weight: bold;
  }
  .highlight-buttons {
    @extend .shadow-2;
    background-color: initial;
    position: absolute !important;
    margin-top: 20px;
    z-index: 12345;
    .highlighter-button,
    .highlight-comment-button {
      @extend .black, .radius-1, .shadow-2;
      padding: 5px 20px;
      font-weight: bold;
      color: $white;
    }
  }
  .highlight-comment-modal {
    .modal-content {
      background: initial;
      border: 0;
      box-shadow: none;
      .modal-body {
        background: initial !important;
        padding: 5px;
        .Form {
          @extend .white;
          margin-top: 15px;
        }
      }
    }
  }
}

@media screen and (max-width: 760px) {
  .ArticleHighlight {
    .highlight-comment-modal {
      .modal-content {
        width: 90%;
      }
    }
  }
}
