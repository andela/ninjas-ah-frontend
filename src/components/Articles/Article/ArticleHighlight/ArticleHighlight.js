import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHighlighter, faComment, faTimes } from '@fortawesome/free-solid-svg-icons';
import { Form, TextArea, Button, Alert } from '../../../common';
import './ArticleHighlight.scss';
import ArticleHighlightsDetails from './ArticleHighlightsDetails';
import { clearCreateRateStore } from '../../../../actions';
import { highlightArticle, clearHighlightArticleStore } from '../../../../actions/articles';
import { htmlHelper } from '../../../../helpers';

export class ArticleHighlight extends Component {
  state = {
    comment: '',
    highlightedText: '',
    anchorKey: '',
    startIndex: 0,
    stopIndex: 0,
    selectionRectangle: null,
    errors: {},
    message: '',
    highlight: null,
    isSelectionInBody: false,
    highlightCommentModalStyle: 'none'
  };

  componentWillReceiveProps = (nextProps) => {
    const { errors, message } = nextProps;
    this.setState(prevState => ({ ...prevState, message, errors }));
  };

  componentDidMount = () => {
    window.addEventListener('selectstart', this.checkHighlightedText);
    window.addEventListener('click', this.checkHighlightedText);
    window.addEventListener('resize', this.checkHighlightedText);
  };

  componentWillUnmount = () => {
    const { clearCreateRateStore, clearHighlightArticleStore } = this.props;
    clearCreateRateStore();
    clearHighlightArticleStore();
  };

  hideModal = (modal) => {
    const { clearHighlightArticleStore } = this.props;
    clearHighlightArticleStore();
    this.setState(prevState => ({ ...prevState, comment: '', [modal]: 'none' }));
  };

  showModal = (modal) => {
    const { clearHighlightArticleStore } = this.props;
    clearHighlightArticleStore();
    this.setState(prevState => ({ ...prevState, [modal]: 'block' }));
  };

  getSelectionIndexes = ({ anchorOffset, focusOffset, anchorKey }) => {
    const startIndex = Math.min(anchorOffset, focusOffset);
    const stopIndex = Math.max(anchorOffset, focusOffset);
    return { startIndex, stopIndex, anchorKey };
  };

  getHighlightedText = (articleBody, anchorKey, startIndex, stopIndex) => {
    const articleBodyBlock = JSON.parse(articleBody).blocks.filter(b => b.key === anchorKey)[0];
    const articleBodyText = (articleBodyBlock && articleBodyBlock.text) || '';
    const highlightedText = articleBodyText.substring(startIndex, stopIndex);
    return highlightedText;
  };

  getSelectionRange = (windowSelection) => {
    const selectionRange = windowSelection.toString() && windowSelection.getRangeAt(0);
    const selectionRectangle = selectionRange && selectionRange.getBoundingClientRect();
    return { selectionRange, selectionRectangle };
  };

  checkHighlightedText = () => {
    const [{ article, editorState }, windowSelection] = [this.props, window.getSelection()];
    const editorStateSelection = editorState.getSelection();
    const { startIndex, stopIndex, anchorKey } = this.getSelectionIndexes(editorStateSelection);
    const { selectionRectangle } = this.getSelectionRange(windowSelection);
    const highlightedText = this.getHighlightedText(article.body, anchorKey, startIndex, stopIndex);
    const isSelectionInBody = highlightedText && highlightedText === windowSelection.toString();

    this.setState(prevState => ({
      ...prevState,
      isSelectionInBody,
      anchorKey,
      startIndex,
      stopIndex,
      highlightedText: highlightedText || prevState.highlightedText,
      selectionRectangle,
      errors: {},
      message: ''
    }));
  };

  handleChange = e => this.setState({ comment: e.target.value });

  saveHighlight = (e) => {
    e.preventDefault();
    const {
      article: { slug },
      highlightArticle,
      clearCreateRateStore
    } = this.props;
    const { anchorKey, highlightedText, startIndex, stopIndex, comment } = this.state;

    return (
      highlightedText
      && stopIndex
      && highlightArticle({ slug, anchorKey, highlightedText, startIndex, stopIndex, comment })
      && clearCreateRateStore()
    );
  };

  render() {
    const { article, loading } = this.props;

    const {
      selectionRectangle,
      comment,
      highlightedText,
      errors,
      message,
      isSelectionInBody,
      highlightCommentModalStyle
    } = this.state;

    return (
      <div className="ArticleHighlight">
        <div className="small-screen-4">
          <div className="small-screen-4">
            {!isSelectionInBody
              && (message || errors.token || errors.authentication || errors.message) && (
                <Alert
                  alertType={(message && 'success') || 'danger'}
                  message={message || errors.token || errors.authentication || errors.message}
                />
            )}
          </div>
          {isSelectionInBody ? (
            <div
              className="highlight-buttons"
              style={{
                top: `${selectionRectangle.top + window.scrollY}px`,
                left: `${selectionRectangle.left}px`
              }}
            >
              <Button buttonClass="button highlighter-button" onClick={this.saveHighlight}>
                <FontAwesomeIcon icon={faHighlighter} size="1x" />
              </Button>
              <Button
                buttonClass="button highlight-comment-button"
                onClick={() => this.showModal('highlightCommentModalStyle')}
              >
                <FontAwesomeIcon icon={faComment} size="1x" />
              </Button>
            </div>
          ) : (
            ''
          )}
          {/* Highlight comment modal */}
          <div className={`highlight-comment-modal modal ${highlightCommentModalStyle}`}>
            <div className="modal-content">
              <div className="modal-header left-align">
                <Button
                  buttonClass="button close-highlight-comment-modal medium-padding yellow radius-5 text-black"
                  onClick={() => this.hideModal('highlightCommentModalStyle')}
                >
                  <FontAwesomeIcon icon={faTimes} size="2x" />
                </Button>
              </div>
              <div className="modal-body">
                <Form onSubmit={this.saveHighlight}>
                  <TextArea
                    name="highlight-comment"
                    textAreaClass="radius-5 medium-text textarea"
                    value={comment}
                    rows={4}
                    onChange={this.handleChange}
                    isRequired={true}
                  />
                  <div className="center-align">
                    <Button type="submit" loading={loading}>
                      Comment
                    </Button>
                  </div>
                </Form>
                <div className="divider" />
                <div className="selection">"{highlightedText}"</div>
              </div>
            </div>
            {htmlHelper.tagGenerator('br', null, 10)}
          </div>
          {/* Highlight details modal */}
          {(article.highlights && article.highlights.length && (
            <ArticleHighlightsDetails article={article} />
          ))
            || ''}
        </div>
      </div>
    );
  }
}

ArticleHighlight.propTypes = {
  article: PropTypes.object,
  loading: PropTypes.bool,
  message: PropTypes.string,
  errors: PropTypes.object,
  editorState: PropTypes.object,
  getSelection: PropTypes.func,
  highlightArticle: PropTypes.func,
  clearHighlightArticleStore: PropTypes.func,
  clearCreateRateStore: PropTypes.func
};

const mapStateToProps = ({ articles: { highlight: { loading, message, errors } } }) => ({
  loading,
  message,
  errors
});

export default connect(
  mapStateToProps,
  { highlightArticle, clearHighlightArticleStore, clearCreateRateStore }
)(ArticleHighlight);
