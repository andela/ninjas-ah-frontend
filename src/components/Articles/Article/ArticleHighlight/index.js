import ArticleHighlight from './ArticleHighlight';

export default ArticleHighlight;
