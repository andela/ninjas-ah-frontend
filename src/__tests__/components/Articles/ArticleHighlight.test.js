import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { EditorState, convertFromRaw } from 'draft-js';
import '../../../__mocks__/window';
import { article, newHighlight } from '../../../__mocks__/article';
import user from '../../../__mocks__/user';
import store from '../../../__mocks__/store';
import ArticleHighlight, { ArticleHighlight as ArticleHighlightComponent } from '../../../components/Articles/Article/ArticleHighlight/ArticleHighlight';
import { mount, shallow } from '../../../../config/enzymeConfig';

const state = {
  comment: 'comment',
  anchorKey: 'cnu26',
  highlightedText: '',
  startIndex: 0,
  stopIndex: 10,
  selectionRectangle: { top: 0, bottom: 0, left: 0, right: 0 },
  errors: {},
  message: '',
  highlight: null,
  isSelectionInBody: true,
  highlightCommentModalStyle: 'none'
};

const props = {
  profile: { ...user },
  editorState: EditorState.createWithContent(convertFromRaw(JSON.parse(article.body))),
  article: { ...article, highlights: [newHighlight] },
  match: { params: { slug: 'slug-slug-slug' } },
  highlightArticle: jest.fn(() => true),
  clearCreateRateStore: jest.fn(() => true),
  clearHighlightArticleStore: jest.fn(() => true),
  message: '',
  errors: { message: 'Network error' }
};

describe('<ArticleHighlightComponent />', () => {
  it('should render without crashing', () => {
    const component = mount(<Provider store={store}>
        <MemoryRouter>
          <ArticleHighlight {...props} />
        </MemoryRouter>
      </Provider>);
    expect(component).toHaveLength(1);
  });
  it('should render without crashing', () => {
    const component = mount(<ArticleHighlightComponent {...{ ...props, article: { ...article, highlights: [] } }} />);
    expect(component).toHaveLength(1);
  });
  it('should render without crashing', () => {
    const component = shallow(<ArticleHighlightComponent {...{ ...props, article: { ...article, highlights: [] } }} />);
    component.instance().componentWillUnmount();
    expect(component).toHaveLength(1);
  });

  it('should show a model containing a textarea to comment on a highlighted text', () => {
    const component = shallow(<ArticleHighlightComponent {...props} />);
    component.setProps(props);
    component.setState(state);

    const highlightCommentForm = component.find('.highlight-comment-modal Form');
    const highlightComment = component.find('TextArea[name="highlight-comment"]');
    const highlightCommentButton = component.find('[buttonClass*="highlight-comment-button"]');
    const closeHighlightCommentButton = component.find('[buttonClass*="close-highlight-comment-modal"]');

    highlightCommentButton.simulate('click', {});
    highlightComment.simulate('change', { target: { value: 'comment' } });
    highlightCommentForm.simulate('submit', { preventDefault: jest.fn() });

    expect(component.state().highlightCommentModalStyle).toEqual('block');
    closeHighlightCommentButton.simulate('click', {});
    expect(component.state().highlightCommentModalStyle).toEqual('none');
    expect(component).toHaveLength(1);
  });

  it('should return a highlighted text', () => {
    const component = shallow(<ArticleHighlightComponent {...props} />);
    const highlightedText = component
      .instance()
      .getHighlightedText(props.article.body, state.anchorKey, 0, 10);
    expect(highlightedText.length).toEqual(10);
    expect(component).toHaveLength(1);
  });

  it('should return the starting index and stopping index of the highlighted text', () => {
    const component = shallow(<ArticleHighlightComponent {...props} />);
    const indexes = component.instance().getSelectionIndexes(props.editorState);
    expect(indexes).toHaveProperty('startIndex');
    expect(indexes).toHaveProperty('stopIndex');
    expect(component).toHaveLength(1);
  });

  it('should return the selection range of the highlighted text', () => {
    const component = shallow(<ArticleHighlightComponent {...props} />);
    const selectionRange = component.instance().getSelectionRange(window.getSelection());
    expect(selectionRange).toHaveProperty('selectionRange');
    expect(selectionRange).toHaveProperty('selectionRectangle');
    expect(component).toHaveLength(1);
  });

  it('should check and update the state when a text is highlighted', () => {
    const component = shallow(<ArticleHighlightComponent {...props} />);
    component.instance().checkHighlightedText();
    expect(component.state()).toHaveProperty('startIndex');
    expect(component.state()).toHaveProperty('stopIndex');
    expect(component.state().selectionRectangle).toHaveProperty('top');
    expect(component.state().selectionRectangle).toHaveProperty('bottom');
    expect(component.state().selectionRectangle).toHaveProperty('left');
    expect(component.state().selectionRectangle).toHaveProperty('right');
    expect(component).toHaveLength(1);
  });
});
