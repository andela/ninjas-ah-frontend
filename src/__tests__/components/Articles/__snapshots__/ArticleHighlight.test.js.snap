// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleHighlightComponent /> should check and update the state when a text is highlighted 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightComponent /> should render without crashing 1`] = `ReactWrapper {}`;

exports[`<ArticleHighlightComponent /> should render without crashing 2`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightComponent /> should return a highlighted text 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightComponent /> should return the selection range of the highlighted text 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightComponent /> should return the starting index and stopping index of the highlighted text 1`] = `ShallowWrapper {}`;

exports[`<ArticleHighlightComponent /> should show a model containing a textarea to comment on a highlighted text 1`] = `ShallowWrapper {}`;
